### PR TITLE
bump CNI to v0.6.0

### LINKED ISSUE #49480: Use CNI v0.6.x in Kubernetes v1.9.0
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

It's time to decide which CNI version to use for v1.8.
We should do early in the cycle to catch regressions in the CNI <-> k8s integration.
We don't want the thing that happened in v1.6 to occur again, so we need to be very careful.

Currently, we're using v0.5.1 and consuming tarballs via a sha pushed by this `Makefile`: https://github.com/kubernetes/kubernetes/blob/master/build/cni/Makefile

We did this earlier because CNI didn't provide binary artifacts for all platform that kubernetes supports.
Since v0.5.2, they do, so we can get rid of `build/cni/Makefile` :tada:

Also, we don't want to consume a strange SHA as the version indicator, instead we want to use a version number _and_ possibly a SHA for verification.

I guess it would make sense to bump CNI to v0.5.2 for the v1.8 cycle.
Agree/disagree? It's not clear to me when v0.6.0 will be released, but the rc has been out for some time already...

cc @kubernetes/sig-network-feature-requests @ixdy @freehan @kubernetes/sig-cluster-lifecycle-feature-requests @dcbw

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #49480

**Special notes for your reviewer**:
/assign @luxas @bboreham @feiskyer 

**Release note**:

```release-note
bump CNI to v0.6.0
```

### CODE REVIEW

**Comment by dnardo** [comment]
cc @danehans

**Comment by freehan** [comment]
Is `https://github.com/containernetworking/cni/releases/download/` reliable enough to take bombardment for every k8s node bootstrap?

**Comment by dixudx** [comment]
> Is https://github.com/containernetworking/cni/releases/download/ reliable enough to take bombardment for every k8s node bootstrap?

Aha, this is also a critical issue. But the package is just about 17MB. It won't occupy too much bandwidth. It is easily downloaded in several seconds.

File: build/debian-hyperkube-base/Makefile
Comment (squeed) [inline build/debian-hyperkube-base/Makefile]: You almost certainly want to download the plugins release instead. See https://github.com/containernetworking/plugins/releases

**Review by squeed — CHANGES_REQUESTED** [review]

**Comment by squeed** [comment]
Watch out, the CNI plugins (e.g. bridge, ptp) moved to a new repository. You need to download from https://github.com/containernetworking/plugins/releases

**Comment by freehan** [comment]
@danehans Do you need this in 1.8? or can you live with it in HEAD (after 1.8 deadline)?

I asked this because the problems of libcni and cni plugins usually take much longer to expose. And they are a pain to fix. (fix push to cni, revendor cni, cherry-pick, cut k8s release).

**Comment by luxas** [comment]
@freehan please see https://github.com/kubernetes/kubernetes/issues/49480#issuecomment-324971090

**Comment by danehans** [comment]
@freehan this PR is needed for IPv6 support in 1.9.

@dixudx can you rebase and get the tests to pass?

/assign @thockin

**Comment by dixudx** [comment]
ping @luxas @danehans @freehan PTAL. Need your `approval` to get merged.

**Review by squeed — APPROVED** [review]
As a CNI maintainer: :+1:

**Comment by thockin** [comment]
I can't review the code changes here.  Assign back to me once it is LGTM'ed and ready for approval.

**Review by feiskyer — APPROVED** [review]
LGTM

File: cluster/gce/gci/configure.sh
Comment (bowei) [inline cluster/gce/gci/configure.sh]: This has to be downloaded from GCS not Github.

@freehan -- let's upload a CNI 0.60 tarball to GCS.

File: cluster/gce/container-linux/configure.sh
Comment (bowei) [inline cluster/gce/container-linux/configure.sh]: See comment below about GCS

File: cluster/gce/gci/configure.sh
Comment (freehan) [reply cluster/gce/gci/configure.sh]: I will take this up this week.

Comment (luxas) [reply cluster/gce/gci/configure.sh]: cc @kubernetes/sig-release-feature-requests @timothysc @jdumars 
Should this really be downloaded from GCS? We should make sure the content there is transparently pushed / handled, find a process for it, make it not Googler-only pushable, make sure checksums are used for verification, etc.

I don't understand why we need to push this to GCS anymore? Earlier we had to do it for a) we used a "random" git sha and b) release binaries were missing for some architectures.

Those two issues are now fixed. I think we should use Github if just possible.

**Review by luxas — APPROVED** [review]
/lgtm

**Comment by luxas** [comment]
We need a **huge release note** for this. A general PSA from someone on kubernetes-dev, Twitter etc. would be helpful.
kubernetes/release debs/rpms need to be updated.
We need to do outreach to the different CNI plugins (Weave, Calico, Canal) that we upgraded the version.
We need to coordinate with @kubernetes/sig-node-feature-requests to make sure the CRI code is ready for this.
We need to coordinate with the well-known CRI implementations (cri-containerd, cri-o, frakti, etc.) and make sure their v1.9-release-targeted version uses the same CNI spec, etc.

This is not simply a code effort; it is a cross-SIG and cross-project wide effort. Last time this caused issues. Let's make sure that doesn't happen again by communicating this change to the broader community. Thank you all!

**Comment by timothysc** [comment]
PSA - @caseydavenport

File: cluster/gce/gci/configure.sh
Comment (bowei) [reply cluster/gce/gci/configure.sh]: The CNI download here is run only for the GCE environment, it would be better to avoid cross dependencies with Github and depend only on GCE itself (i.e. GCS). For example, if Github went down, then it would take out GCE-based clusters.

Comment (luxas) [reply cluster/gce/gci/configure.sh]: Ok, earlier the GCS builds have been used everywhere else as well. Anyway, let's just move the Github-released artifacts then, not reinvent the releasing scripts and build things yourselves.
It should be a 1:1 mirror of Github then.

Comment (luxas) [reply cluster/gce/gci/configure.sh]: And provide visibility to others for this process.

Comment (freehan) [reply cluster/gce/gci/configure.sh]: Uploaded CNI tarball to GCS:
link: 
https://storage.googleapis.com/kubernetes-release/network-plugins/cni-plugins-amd64-v0.6.0.tgz

**Comment by freehan** [comment]
Please point to the GCS. I will lgtm it.

File: cluster/gce/gci/configure.sh
Comment (bowei) [reply cluster/gce/gci/configure.sh]: I think your suggestion that will depend on a change in the CNI repository? We can move ahead with this fix in the mean time while that is being resolved?

Comment (dixudx) [reply cluster/gce/gci/configure.sh]: > I don't understand why we need to push this to GCS anymore? Earlier we had to do it for a) we used a "random" git sha and b) release binaries were missing for some architectures.

> Those two issues are now fixed. I think we should use Github if just possible.

@luxas Agree. No need to have another copy for GCS. @bowei We did this earlier because CNI didn't provide binary artifacts for all platforms that kubernetes supports. So we build arch-specific artifacts. Since v0.5.2, they do.

Comment (bowei) [reply cluster/gce/gci/configure.sh]: `cluster/gce/**` is only applicable to GCE environment. I think generically your comment is true, but not for the GCE-related cluster configuration scripts.

**Comment by squeed** [comment]
Hi there, @luxas  - I'm not sure that this is that big of a change. There are two parts to this PR: bumping the bundled plugins and bumping the "libcni" vendored in to the kubelet. 

There are no notable libcni changes in this release, just a few bug fixes. The big libcni change was the bump to cni v0.5, which is already in k8s 1.7. However, we've worked hard to ensure libcni backwards compatible (and it does version down- and up-conversion), so nothing serious here.

As for the plugins, this is a bigger change - the previously released versions were from April. However, there's not really much that's different; they now support ipv6, but there aren't any other big changes, and they're all backwards compatible. Plus, vendors such as weave and calico use their own plugins, so they won't be affected.

I've been meaning to write a blogpost explaining the value of chained plugins and how to use portmapping,  but this is a new feature, not a breaking change.

**Comment by dixudx** [comment]
@luxas @freehan @squeed Updated the download link to GCS. PTAL. Thanks.

**Comment by luxas** [comment]
@squeed I hope you're not opposed to communicating broad changes clearly.
What I said was not a technical issue. Last time (v1.6 cycle) we bumped CNI, and various side-effects were discovered in different parts of the project; which for example made kubeadm v1.6.0 unusable; as the kubelet behavior related to networking changed remarkably really late in the cycle.

I'm not criticizing nor blaming anyone, I'm just informing that although nothing _should_ change, there are so many various users of CNI in different places, so use-cases / combinations / edge cases nobody could ever think of might hit us.

As nobody of us want that to happen, we need to do this early in the cycle (which we are doing -- very good), but also be super-clear with the communication aspect of the matter.

Thank you all, let's get this in soon-ish and send some PSAs

**Comment by freehan** [comment]
@dixudx Can you point everything to GCS? we do not want to have another dependency on github. GCS is much more reliable.

**Comment by dixudx** [comment]
ping @freehan Updated all to GCS. PTAL. Thanks.

**Comment by freehan** [comment]
/lgtm

Thank you!

File: build/root/WORKSPACE
Comment (chrislovecnm) [inline build/root/WORKSPACE]: Will this get us a sha file as well?

**Review by chrislovecnm — COMMENTED** [review]
Question

File: build/root/WORKSPACE
Comment (dixudx) [reply build/root/WORKSPACE]: @freehan Have you updated the sha file?

Comment (freehan) [reply build/root/WORKSPACE]: ah...nope. Why do you need the sha file?

Comment (dixudx) [reply build/root/WORKSPACE]: ack @chrislovecnm 

I think we only need to know the sha value, which is exactly the same with that of [cni-plugins github releases](https://github.com/containernetworking/plugins/releases), since the tarball stays unchanged, right?

If you have other requirements, @freehan can help upload it.

Comment (luxas) [reply build/root/WORKSPACE]: Yes, @freehan should just move over the shasums

Comment (chrislovecnm) [reply build/root/WORKSPACE]: We have sha files for every other tar package with the k/k release. But no sha file for this file. That is not great consistency with our release. Does bazel support it? Does k/k need it no, but systems that consume cni need it.

Comment (luxas) [reply build/root/WORKSPACE]: @chrislovecnm FWIW, I think the GCS thing here is just about storing the exact bins/shasums listed in https://github.com/containernetworking/cni/releases on GCS instead of Github. The shasums should also be moved over

Comment (freehan) [reply build/root/WORKSPACE]: Ack. I will upload the sha files

File: build/debian-hyperkube-base/Makefile
Comment (ixdy) [inline build/debian-hyperkube-base/Makefile]: we should have bumped the `debian-hyperkube-base` tag here.

**Comment by ixdy** [comment]
Who built/pushed the new CNI tarballs to gs://kubernetes-release/network-plugins? We're missing s390x and ppc64le.

**Comment by dixudx** [comment]
@freehan Please help uploading those missing files as well. Thanks.

File: build/debian-hyperkube-base/Makefile
Comment (dixudx) [reply build/debian-hyperkube-base/Makefile]: Will do in another PR.

**Comment by ixdy** [comment]
I realized the tarballs are just copied from https://github.com/containernetworking/plugins/releases, so I copied over the ppc64le and s390x ones too.

**Comment by bowei** [comment]
@ixdy can you also copy the .sha files?

**Comment by freehan** [comment]
I just uploaded the sha files. Thanks!

**Comment by luxas** [comment]
Great! Thank you @freehan and @ixdy :+1:!

**Comment by chrislovecnm** [comment]
/reopen

I am can d/l

https://storage.googleapis.com/kubernetes-release/network-plugins/cni-plugins-amd64-v0.6.0.tgz.sha1

but I am unable to d/l

https://storage.googleapis.com/kubernetes-release/network-plugins/cni-plugins-amd64-v0.5.0.tgz.sha1

Or the tarball.  Can we get 0.4.0 and 0.5.0 uploaded with there SHAs?  I am uncertain which versions are for k8s 1.5 and k8s 1.6, as I only have the names with the shas, and the tarball that is staged does not have the version in it.

To be clear I am looking for these two previous files that should have new names and sha files.

https://storage.googleapis.com/kubernetes-release/network-plugins/cni-07a8a28637e97b22eb8dfe710eeae1344f69d16e.tar.gz
https://storage.googleapis.com/kubernetes-release/network-plugins/cni-0799f5732f2a11b329d9e3d51b9c8f2e3759f2ff.tar.gz"


cc @freehan @luxas 

Thanks

**Comment by dixudx** [comment]
> I am not certain who copied the 0.6.0 tar ball over, but I would love to have the previous two and there sha1 files as well.

@chrislovecnm Please contact @freehan if needed.

**Comment by chrislovecnm** [comment]
@dixudx I filed #57341 so we can stop talking on a closed issue

**Pull request merged**
